state/allwatcher_internal_test.go

```diff
@@ -4046,6 +4072,88 @@ func testChangeApplicationOffers(c *gc.C, runChangeTests func(*gc.C, []changeTes
 	runChangeTests(c, changeTestFuncs)
 }
 
+func testChangeGenerations(c *gc.C, runChangeTests func(*gc.C, []changeTestFunc)) {
+	changeTestFuncs := []changeTestFunc{
+		func(c *gc.C, st *State) changeTestCase {
+			return changeTestCase{
+				about: "no change if generation absent from state and store",
+				change: watcher.Change{
+					C:  "generations",
+					Id: st.docID("does-not-exist"),
+				}}
+		},
+		func(c *gc.C, st *State) changeTestCase {
+			return changeTestCase{
+				about: "generation is removed if not in backing store",
+				initialContents: []multiwatcher.EntityInfo{
+					&multiwatcher.GenerationInfo{
+						ModelUUID: st.ModelUUID(),
+						Id:        "to-be-removed",
+					},
+				},
+				change: watcher.Change{
+					C:  "generations",
+					Id: st.docID("to-be-removed"),
+				},
+			}
+		},
+		func(c *gc.C, st *State) changeTestCase {
+			c.Assert(st.AddBranch("new-branch", "some-user"), jc.ErrorIsNil)
+			branch, err := st.Branch("new-branch")
+			c.Assert(err, jc.ErrorIsNil)
+
+			return changeTestCase{
+				about: "generation is added if in backing but not in tore",
+				change: watcher.Change{
+					C:  "generations",
+					Id: st.docID(branch.doc.DocId),
+				},
+				expectContents: []multiwatcher.EntityInfo{
+					&multiwatcher.GenerationInfo{
+						ModelUUID:     st.ModelUUID(),
+						Id:            st.localID(branch.doc.DocId),
+						Name:          "new-branch",
+						AssignedUnits: map[string][]string{},
+					}},
+			}
+		},
+		func(c *gc.C, st *State) changeTestCase {
+			c.Assert(st.AddBranch("new-branch", "some-user"), jc.ErrorIsNil)
+			branch, err := st.Branch("new-branch")
+			c.Assert(err, jc.ErrorIsNil)
+
+			app := AddTestingApplication(c, st, "wordpress", AddTestingCharm(c, st, "wordpress"))
+			u, err := app.AddUnit(AddUnitParams{})
+			c.Assert(err, jc.ErrorIsNil)
+
+			c.Assert(branch.AssignUnit(u.Name()), jc.ErrorIsNil)
+
+			return changeTestCase{
+				about: "generation is updated if in backing and in store",
+				change: watcher.Change{
+					C:  "generations",
+					Id: st.docID(branch.doc.DocId),
+				},
+				initialContents: []multiwatcher.EntityInfo{
```

Comment:
I found this test a little confusing, becuase the `AddBranch` above is, I think, unrelated to this definition in the initialContents. It seems that the initial code above needs to create the branch in order to be able to have a unit track it, but the intialContents is here to illustrate that it is edited. However the initial contents looks like it would have been created by the `AddBranch` anyway. Am I right in this?

Reply:
Yes, the initial contents is a new branch and the set-up creates a modified representation in state. Sending the change updates the store with the modifications.